### PR TITLE
[WIP] Megatron Model Parallelism in NeMo

### PR DESCRIPTION
**Pull request by ericharper**
NeMo should have model parallelism because state of the art NLP models often require more model capacity than can fit on a single GPU.

In this PR we begin to add model parallelism to NeMo by first supporting finetuning with the Megatron BERT Neural Module.

Integrating model parallel Megatron models into NeMo requires a little bit of care as Megatron-LM is built from the ground up with model + data parallelism.
Megatron-LM always assumes model parallelism even if it is the trivial case of model parallel size 1.

NeMo's approach to distributed PyTorch assumes data parallelism only. 
After integrating model parallel Megatron models into NeMo, NeMo will have a more robust and intuitive approach to distributed PyTorch.

# TODO
- [x] Update NeMo so that model parallel Megatron models can be trained easily.
  - [x] User only needs to specify model_parallel_size flag in neural factory
  - [ ] Callback works automatically for model parallel
  - [x] Training works automatically for model parallel
  - [x] Eval works automatically for model parallel
  - [ ] Infer works automatically for model parallel
- [x] Validate that Model Parallel Megatron BERT models trained with NeMo achieve similar accuracies to those trained with Megatron-LM
  - [x] GLUE
    - [x] MNLI
    - [x] QQP
  - [x] SQUAD
    - [x] 1.1
    - [x] 2.0
  - [ ] RACE (not in NeMo NLP Examples)
- [ ] Validate scaling performance
  - [x] High GPU Utilization during training
  - [ ] Near linear data parallel scaling

# Changes needed for Megatron Style Model Parallel Training
- [x] Initialization
  - [x] Model parallel intialization should happen when instantiating the factory
  - [x] Model parallel rank
  - [x] Model parallel size
  - [x] Model parallel group
  - [x] Data Parallel rank
  - [x] Data parallel size
  - [x] Data parallel group
- [ ] Checkpointing
  - [x] Restore checkpoints from Megatron-LM
  - [ ] Models now have checkpoints for each model parallel rank
  - [ ] Directory should be given to restore checkpoints from
  - [ ] Checkpoint callback should save and restore mp checkpoints
  - [ ] Only data parallel rank 0 should be saving checkpoints
- [ ] Training
  - [x] Distributed data parallel (DDP)
    - [x] process_group is the data_parallel_group
  - [x] Distributed Sampler
    - [x] rank is data_parallel_rank
    - [x] num_replicas is data_parallel_size
  - [ ] Data Broadcast (for pretraining only)
    - [ ] Data should only be broadcast from CPU to model parallel rank 0
  - [x] Gradient clipping - reduce grads across model parallel group
  - [x] Apex Amp Loss Scaling - reduce overflow across model parallel group
  - [ ] Weight Decay - Megatron-LM does not decay layer norms and biases

# Changes Made

## **nemo/core**
neural_factory.py
- Update model parallel rank, size, and data parallel rank, size, group properties in AppState() after torch distributed initialization
- Add model parallel size argument to NeuralModuleFactory
- Add random_seed argument to NeuralModuleFactory and update AppState()
- Add megatron model parallel initialization to NeuralModuleFactory

## **nemo/utils**
app_state.py
- Add model parallel rank, size, and data parallel rank, size, group properties in AppState()

## **nemo/backends/pytorch**
actions.py
- Added data_parallel_rank/size/group and model_parallel_rank to PtActions
- Update DDP proces_group to data_parallel_group
- Update distributed sampler for _train, _eval to use rank is data_parallel_rank and num_replicas is data_parallel_size
- Add mpu gradient clipping
- Add monkey patch for apex amp loss scaling

## **collections/nlp**
megatron_bert_nm.py
- Remove initialize_megatron call since model and data parallel groups are being initialized from the factory
- Add remaining initializations needed for megatron-lm
- Add megatron_restore_from to load pre-trained models coming from megatron_lm

common_utils.py
- Add megatron_restore_from if megatron checkpoint

## **examples/nlp**
token_classification.py,
glue_benchmark_with_bert.py,
question_answering_squad.py
- Update to work with model parallel megatron

### CODE REVIEW

File: nemo/collections/nlp/nm/trainables/common/common_utils.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def get_pretrained_lm_models_list():
     return get_megatron_lm_models_list() + get_huggingface_lm_models_list()
 
 
-def get_pretrained_lm_model(pretrained_model_name, config=None, vocab=None, checkpoint=None):
+def get_pretrained_lm_model(pretrained_model_name, config=None, vocab=None, checkpoint=None, local_rank=None):
```
Comment (ekmb) [inline nemo/collections/nlp/nm/trainables/common/common_utils.py:34]: could you please update the docstring with the new arg

File: nemo/collections/nlp/nm/trainables/common/common_utils.py [hunk 2]
```diff
@@ -70,6 +70,9 @@ def get_pretrained_lm_model(pretrained_model_name, config=None, vocab=None, chec
         raise ValueError(f'{pretrained_model_name} is not supported')
 
     if checkpoint:
-        model.restore_from(checkpoint)
+        if pretrained_model_name == 'megatron-bert-cased' or pretrained_model_name == 'megatron-bert-uncased':
+            model.megatron_restore_from(checkpoint, local_rank)
```
Comment (ekmb) [inline nemo/collections/nlp/nm/trainables/common/common_utils.py:74]: this should be `if 'megatron' in pretrained_model_name:`

File: nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py [hunk 3]
```diff
@@ -124,8 +130,18 @@ def forward(self, input_ids, attention_mask, token_type_ids):
             input_ids, position_ids, extended_attention_mask, tokentype_ids=token_type_ids
         )
         return sequence_output
-
-    def restore_from(self, path, local_rank=0):
+    
+    def megatron_restore_from(self, path, local_rank=None):
```
Comment (ekmb) [inline nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py:134]: maybe rename this one to smth like restore_model_parallel_megatron? also check that the new path exists? 
so for this to work, users should provide path to the checkpoint folder not the checkpoint .pt file, right? we need to make it clear in the training scripts args

File: nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py [hunk 4]
```diff
@@ -90,7 +90,13 @@ def __init__(
             "vocab_file": vocab_file,
         }
 
-        initialize_megatron(None, megatron_args, ignore_unknown_args=True)
+        set_global_variables(extra_args_provider=None,
+                args_defaults=megatron_args,
+                ignore_unknown_args=True)
+        if self.factory._random_seed is None:
+            raise ValueError("Megatron Neural Module requires Neural Factory to have random_seed is not None.")
```
Comment (ekmb) [inline nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py:97]: should we maybe throw a warning but set self.factory._random_seed to some value and keep going?

File: nemo/core/neural_factory.py [hunk 5]
```diff
@@ -394,6 +427,14 @@ def torch_broadcast_wrapper(str_len=None, string=None, src=0):
                     return return_string
 
                 broadcast_func = torch_broadcast_wrapper
+
+                if self._model_parallel_size is not None:
+                    from megatron import mpu
+                    mpu.initialize.initialize_model_parallel(self._model_parallel_size)
+                    self._mp_rank = mpu.get_model_parallel_rank()
```
Comment (ekmb) [inline nemo/core/neural_factory.py:434]: Megatron init wouldn't work from MegatronNM?

File: nemo/collections/nlp/nm/trainables/common/common_utils.py (hunk 1 above)
Comment (ericharper) [reply nemo/collections/nlp/nm/trainables/common/common_utils.py]: Done.

File: nemo/collections/nlp/nm/trainables/common/common_utils.py (hunk 2 above)
Comment (ericharper) [reply nemo/collections/nlp/nm/trainables/common/common_utils.py:81]: Done.

File: nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py (hunk 3 above)
Comment (ericharper) [reply nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py:134]: Yes, this is how I updated the checkpoint arg for token_classificaction.py:
`parser.add_argument(
    "--bert_checkpoint",
    default=None,
    type=str,
    help="Path to model file. (If using model parallel, path to checkpoint directory.)"
)`

Comment (ericharper) [reply nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py]: Updated the naming, and checking paths.

File: nemo/core/neural_factory.py (hunk 5 above)
Comment (ericharper) [reply nemo/core/neural_factory.py:430]: It would work but the idea of this PR is to initialize torch.distributed groups with the factory and then store world information like local_rank, global_rank, mp_rank, dp_rank as properties of the factory.  Note this already being done for data parallel. This has the benefit of giving easier access to world information to NeMo users.

To me this is also an advantage of NeMo's abstraction. If we build up the model parallel API enough, users could potentially use NeMo as a framework for model parallel architecture development. 

If we just call mpu.initialize_megatron from the neural module, the communication information is stored as global variables and somewhat hidden from NeMo users.

File: nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py (hunk 4 above)
Comment (ericharper) [reply nemo/collections/nlp/nm/trainables/common/megatron/megatron_bert_nm.py]: Done.

File: nemo/core/neural_factory.py (hunk 5 above)
Comment (ericharper) [reply nemo/core/neural_factory.py]: After some more investigation, I don't think we can initialize megatron from the megatron neural module when model parallel size is greater than one. This is because the training loop is instantiated (ptActions) when the factory is initialized. In particular, DDP and DistributedSampler arguments depend on the data parallel groups.

**Comment by okuchaiev** [comment]
@ericharper what should we do about this PR?

**Comment by ericharper** [comment]
Closing since model parallelism is being implemented in new NeMo.
